File: Search Insert Position - Leetcode 35/Search Insert Position - Leetcode 35.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        n = len(nums)
        l = 0
        r = n - 1

        while l <= r:
            m = (l + r) // 2
            
            if nums[m] < target:
                l = m + 1
            elif nums[m] > target:
                r = m - 1
            else:
                return m

        if nums[m] < target:
            return m + 1
        else:
            return m

# Time Complexity: O(log n)
# Space Complexity: O(1)

File: Search Insert Position - Leetcode 35/Search Insert Position - Leetcode 35.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        while (l <= r) {
            int m = (l + r) / 2;

            if (nums[m] < target) {
                l = m + 1;
            } else if (nums[m] > target) {
                r = m - 1;
            } else {
                return m;
            }
        }

        return l;
    }
};
